### PR TITLE
[SPARK-11615] Drop @VisibleForTesting annotation

### PR DESCRIPTION
**Pull request by tedyu**
See http://search-hadoop.com/m/q3RTtjpe8r1iRbTj2 for discussion.

Summary: addition of @VisibleForTesting annotation resulted in spark-shell malfunctioning.

### CODE REVIEW

**Comment by tedyu** [comment]
May need a bit hint on how style-checker rule should be composed to prevent regression

**Comment by tedyu** [comment]
Here is listing of files where @VisibleForTesting is used:
```
import com.google.common.annotations.VisibleForTesting
  @VisibleForTesting
./core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala
import com.google.common.annotations.VisibleForTesting
./core/src/main/scala/org/apache/spark/ui/jobs/JobProgressListener.scala
import com.google.common.annotations.VisibleForTesting
  @VisibleForTesting
  @VisibleForTesting
./core/src/main/scala/org/apache/spark/util/AsynchronousListenerBus.scala
import com.google.common.annotations.VisibleForTesting
  @VisibleForTesting
./core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala
import com.google.common.annotations.VisibleForTesting
  @VisibleForTesting
./sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala
import com.google.common.annotations.VisibleForTesting
./yarn/src/test/scala/org/apache/spark/network/shuffle/ShuffleTestAccessor.scala
```
Do we want to remove all occurrences ?

**Comment by marmbrus** [comment]
Can you create a JIRA please.

@JoshRosen for hints on the style checker.

**Comment by tedyu** [comment]
I can follow classforname rule from scalastyle-config.xml - if we decide that @VisibleForTesting should not be used at all

**Comment by JoshRosen** [comment]
What I'd probably do is just bad the import via a string pattern match checker, then replace the existing occurrences with comments or introduce a Spark-specific version of the annotation and use that instead.

**Comment by tedyu** [comment]
Added a rule in scalastyle-config.xml

**Comment by tedyu** [comment]
@JoshRosen
Can you take a look ?

File: scalastyle-config.xml [hunk 1]
```diff
@@ -150,6 +150,13 @@ This file is divided into 3 sections:
       // scalastyle:on println]]></customMessage>
   </check>
 
+  <check customId="visiblefortesting" level="error" class="org.scalastyle.file.RegexChecker" enabled="true">
+    <parameters><parameter name="regex">@VisibleForTesting</parameter></parameters>
+    <customMessage><![CDATA[
+			Are you sure that you want to use @VisibleForTesting ? In most cases, you should use // VisibleForTesting instead.
```
Comment (andrewor14) [inline scalastyle-config.xml]: It's weird to have `// VisibleForTesting` as a comment. I'd say we just put it in the java doc of the method, e.g.
```
/**
 * This method does something...
 * Exposed for testing.
 */
```
For the custom message here I'd just say the annotation is banned or something.

**Comment by tedyu** [comment]
From https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45529/consoleFull :
```
DONE 
Tests passed.
```

**Comment by tedyu** [comment]
From https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45532/console :
{code}
Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --progress https://github.com/apache/spark.git +refs/pull/9585/*:refs/remotes/origin/pr/9585/*" returned status code 143:
stdout: 
stderr: error: RPC failed; result=18, HTTP code = 200
{code}

File: scalastyle-config.xml [hunk 2]
```diff
@@ -150,6 +150,14 @@ This file is divided into 3 sections:
       // scalastyle:on println]]></customMessage>
   </check>
 
+  <check customId="visiblefortesting" level="error" class="org.scalastyle.file.RegexChecker" enabled="true">
+    <parameters><parameter name="regex">@VisibleForTesting</parameter></parameters>
+    <customMessage><![CDATA[
+			@VisibleForTesting is banned due to tendency of making Spark Shell malfunction. In most cases,
```
Comment (vanzin) [inline scalastyle-config.xml]: super nit: indentation is weird.

**Comment by vanzin** [comment]
I kinda like that annotation, much more than a comment in the scaladoc, but well... LGTM. retest this please

File: core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala [hunk 3]
```diff
@@ -193,8 +191,8 @@ private[netty] class Inbox(
 
   def isEmpty: Boolean = inbox.synchronized { messages.isEmpty }
 
-  /** Called when we are dropping a message. Test cases override this to test message dropping. */
-  @VisibleForTesting
+  /** Called when we are dropping a message. Test cases override this to test message dropping.
+   * Exposed for testing. */
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala]: should be java doc, not scala doc

File: scalastyle-config.xml [hunk 4]
```diff
@@ -150,6 +150,14 @@ This file is divided into 3 sections:
       // scalastyle:on println]]></customMessage>
   </check>
 
+  <check customId="visiblefortesting" level="error" class="org.scalastyle.file.RegexChecker" enabled="true">
+    <parameters><parameter name="regex">@VisibleForTesting</parameter></parameters>
+    <customMessage><![CDATA[
+      @VisibleForTesting is banned due to tendency of making Spark Shell malfunction. In most cases,
+      you should describe purpose for testing in scala doc.
```
Comment (andrewor14) [inline scalastyle-config.xml]: I would just say
```
@VisibleForTesting causes classpath issues. Please note this in the java doc instead (SPARK-11615).
```

**Comment by tedyu** [comment]
```
[info] *** 1 TEST FAILED ***
[error] Failed: Total 19, Failed 1, Errors 0, Passed 18, Ignored 1
[error] Failed tests:
[error] 	org.apache.spark.sql.hive.thriftserver.CliSuite
```
Not related to this PR

**Pull request closed without merging**